Add more HTTP tests

davidsh opened the pull request:
* Add more tests ported from the internal ToF tests.
* Cleanup existing tests to be consistent in style and format as it has
drifted due to many different people working on the tests.
* Disposing the HttpClient object will dispose the handler as well so we
don't need multiple 'using' statements unless we are explicitly testing
dispose ordering semantics.

davidsh commented:
cc: @CIPop @stephentoub @SidharthNabar

CIPop commented on src/System.Net.Http.WinHttpHandler/tests/FunctionalTests/ServerCertificateTest.cs:
```diff
@@ -0,0 +1,183 @@
+            foreach (var element in chain.ChainElements)
+            {
+                _validationCallbackHistory.CertificateChain.Add(element.Certificate);
+            }
```
Nit: missing new line.

davidsh replied:
It has the newline at the end of the file.  When it doesn't GitHub will show a red-icon in the code viewer.  But I checked the file and it has a newline at the end.

CIPop replied:
Not at EOF but right here, where the comment is. New lines after closing brace (}) if followed by anything else but another closing brace.

CIPop commented on src/System.Net.Http.WinHttpHandler/tests/FunctionalTests/ServerCertificateTest.cs:
```diff
@@ -0,0 +1,183 @@
+        [Fact]
+        public async Task NoCallback_ValidCertificate_CallbackNotCalled()
+        {
+            var handler = new WinHttpHandler();
```
WinHttpHandler is Disposable. 
A test where it's not properly disposed is interesting. If that's the case, please add a top-of-file general comment explaining that is by intent.

If not, it can easily be included in the using chain:

```C#
using (var handler = new WinHttpHandler())
using (var client = new HttpClient(handler))
{ 
...
}
```

davidsh replied:
It is not necessary to have a `using` for the handler.  There only needs to a `using` for the HttpClient.  Disposing the HttpClient will automatically Dispose the handler inside it.

CIPop replied:
The recommended dispose pattern requires callers of IDisposable objects to ensure Dispose() is called.
Sounds like that's not what you're trying to test. In that case, this applies:

> If that's the case, please add a top-of-file general comment explaining that is by intent.

davidsh replied:
Added comments to HttpClientHandlerTest.cs and WinHttpHandlerTest.cs.

CIPop commented on src/System.Net.Http.WinHttpHandler/tests/FunctionalTests/WinHttpHandlerTest.cs:
```diff
@@ -38,25 +38,18 @@ public void SendAsync_SimpleGet_Success()
         }
 
         [Fact]
-        public async Task GetAsync_Cancel_CancellationTokenPropagates()
```
What is the expected behavior for `GetAsync`?
Could we add the removed test as well?

davidsh replied:
There are a bunch of HTTP tests in System.Net.Http/tests/FunctionalTests that really exercise the functionality.  Since System.Net.Http contains the WinHttpHandler (Windows ) and CurlHandler (*Nix), most of the tests are there.

I'm in the process of reorganizing the HTTP tests.  The intent for System.Net.Http.WinHttpHandler tests is to test specific functionality such as properties that aren't present in the main HttpClientHandler but are present in the WinHttpHandler.

This particular test will probably get moved to HttpClientHandlerTest.cs in the System.Net.Http tests at some point in the future.

CIPop replied:
Agreed. Please add the test to the HttpClient.Functional suite within this PR to make sure we don't forget about it.

CIPop replied:
Got it now: this was renamed not removed.

CIPop commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:
```diff
@@ -317,38 +335,32 @@ public void GetAsync_CredentialIsCredentialCacheUriRedirect_HttpStatusCodeOK()
+                HttpResponseMessage response =
+                    await client.PostAsync(remoteServer, stringContent);
+                Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+                var responseContent = await response.Content.ReadAsStringAsync();
```
Nit: Please name the `var` in cases were the type is not on the same line.

davidsh replied:
fixed.

CIPop commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:
```diff
@@ -317,38 +335,32 @@ public void GetAsync_CredentialIsCredentialCacheUriRedirect_HttpStatusCodeOK()
+                var content = new FormUrlEncodedContent(values);
+                HttpResponseMessage response = await client.PostAsync(remoteServer, content);
+                Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+                var responseContent = await response.Content.ReadAsStringAsync();
```
Nit: name this var (throughout the file).

davidsh replied:
fixed.

CIPop commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:
```diff
@@ -369,25 +381,22 @@ public void GetAsync_CredentialIsCredentialCacheUriRedirect_HttpStatusCodeOK()
-                    }
+                        Name = fileTitle,
+                        FileName = fileName
+                    };
```
Nit missing new line.

davidsh replied:
There is already a newline at the end of the file.

CIPop replied:
Not at EOF: right here, after closing brace.

davidsh replied:
fixed

CIPop commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:
```diff
@@ -369,25 +381,22 @@ public void GetAsync_CredentialIsCredentialCacheUriRedirect_HttpStatusCodeOK()
+                    form.Add(content);
+                    HttpResponseMessage response = await client.PostAsync(remoteServer, form);
+                    Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+                    var responseContent = await response.Content.ReadAsStringAsync();
```
Nit: improper use of `var`.

davidsh replied:
fixed.

CIPop commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:
```diff
@@ -526,30 +529,41 @@ public static IEnumerable<object[]> PostServersStreamsAndExpectedData
+        }
+
+        [Fact]
+        [ActiveIssue("Channel binding not implemented on *Nix", PlatformID.AnyUnix)]
```
I believe we need to put in a GitHub issue instead of a string? (Maybe CI or other systems might break if they see this.)

davidsh replied:
Done. Issue #3565

CIPop commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:
```diff
@@ -560,50 +574,40 @@ public static IEnumerable<object[]> PostServersStreamsAndExpectedData
+                    mediaTypeJson);
+                HttpResponseMessage response = await client.PutAsync(remoteServer, stringContent);
+                Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+                var responseContent = await response.Content.ReadAsStringAsync();
```
Nit: improper use of `var`.

davidsh replied:
fixed.

CIPop commented:
Left a few comments. The one that I care most about is the test that got removed and properly disposing the handler.

davidsh commented:
Addressed PR feedback.  There was no test that was "removed" that I'm aware of.  If not, please help clarify which test you are referring to.

Regarding the Disposing semantics, it's not necessary to explicitly dispose the handler if the client is being disposed.  That was part of the code simplification I did in the test files.

CIPop commented:
LGTM
